### PR TITLE
feat: add eof pdu

### PR DESCRIPTION
**Pull request by Qlas**
#22

### CODE REVIEW

File: include/core/cfdp_core/pdu_directive.hpp [hunk 1]
```diff
@@ -62,4 +62,56 @@ class AckPdu : PduInterface
...
+        const_pdu_size_bytes + sizeof(uint64_t);
+
+    [[nodiscard]] inline uint8_t getSizeOfFileSize() const
+    {
+        return (largeFileFlag == LargeFileFlag::LargeFile) ? sizeof(uint64_t) : sizeof(uint32_t);
+    }
+
+    [[nodiscard]] inline uint8_t getFaultLocationSize() const
+    {
+        return (conditionCode != Condition::NoError) ? sizeof(uint8_t) + lengthOfEntityID : 0;
```
Comment (uncommon-nickname) [inline include/core/cfdp_core/pdu_directive.hpp:113]: you can do that branchless, we would have to check the ASM to see if it is worth it tho

File: src/cfdp_core/pdu_directive.cpp [hunk 2]
```diff
@@ -111,3 +113,119 @@ std::vector<uint8_t> cfdp::pdu::directive::AckPdu::encodeToBytes() const
...
+    {
+        throw exception::DecodeFromBytesException("File Directive code is not End of File Pdu");
+    }
+    const auto secondByte = memory[1];
+
+    conditionCode = Condition((secondByte & eof_condition_code_bitmask) >> 4);
+    checksum      = utils::bytesToInt<uint64_t>(memory, 2, sizeof(uint32_t));
+    fileSize      = utils::bytesToInt<uint64_t>(memory, 6, getSizeOfFileSize());
+
+    if (conditionCode != Condition::NoError)
```
Comment (uncommon-nickname) [inline src/cfdp_core/pdu_directive.cpp:190]: early exit instead of nesting?

File: tests/cfdp_core/pdu_directive.test.cpp [hunk 3]
```diff
@@ -123,43 +137,182 @@ TEST_F(AckPduTest, TestEncodingFinishedAck)
...
 {
     std::array<uint8_t, 3> encoded_frame = {12, 81, 34};
 
-    auto encoded = std::span<uint8_t const>{encoded_frame.begin(), encoded_frame.end() - 1};
+    auto encoded = std::span<uint8_t const>{encoded_frame.begin(), encoded_frame.end()};
+
+    ASSERT_THROW(Ack{encoded}, cfdp::pdu::exception::DecodeFromBytesException);
+}
+
+TEST_F(EndOfFileTest, TestConstructorException)
```
Comment (uncommon-nickname) [inline tests/cfdp_core/pdu_directive.test.cpp:181]: I would parametrize it

**Review by uncommon-nickname — CHANGES_REQUESTED** [review]

File: include/core/cfdp_core/pdu_directive.hpp [hunk 4]
```diff
@@ -62,4 +62,56 @@ class AckPdu : PduInterface
...
+    [[nodiscard]] auto getConditionCode() const { return conditionCode; }
+    [[nodiscard]] auto getFileSize() const { return fileSize; }
+    [[nodiscard]] auto getChecksum() const { return checksum; }
+    [[nodiscard]] auto getLengthOfEntityID() const { return lengthOfEntityID; }
+    [[nodiscard]] auto getFaultEntityID() const { return faultEntityID; }
+
+  private:
+    LargeFileFlag largeFileFlag;
+    Condition conditionCode;
+    uint64_t fileSize;
```
[the remark this answers is not in the document]
Comment (Qlas) [reply include/core/cfdp_core/pdu_directive.hpp:94]: The idea was to force the next level of abstraction to ensure that the fileSize have the right size, without it,  we have to retrieve LargeFileFlag in the constructor and validate that the fileSize is not too large

[the remark this answers is not in the document]
Comment (uncommon-nickname) [reply include/core/cfdp_core/pdu_directive.hpp:94]: Yeah, but this gives you this problem later...

image

[the remark this answers is not in the document]
Comment (uncommon-nickname) [reply include/core/cfdp_core/pdu_directive.hpp:94]: Also, if we take a file size  of `30` and cast it to uint64_t isn't it plainly wrong? This implementation is pretty weird to me

File: include/core/cfdp_core/pdu_directive.hpp [hunk 5]
```diff
@@ -9,12 +9,12 @@ using ::cfdp::pdu::header::LargeFileFlag;
...
 {
 
-class KeepAlivePdu : PduInterface
+class KeepAlive : PduInterface
 {
   public:
-    KeepAlivePdu(uint32_t progress) : progress(progress), largeFileFlag(LargeFileFlag::SmallFile){};
-    KeepAlivePdu(uint64_t progress) : progress(progress), largeFileFlag(LargeFileFlag::LargeFile){};
-    KeepAlivePdu(std::span<uint8_t const> memory);
+    KeepAlive(uint32_t progress) : progress(progress), largeFileFlag(LargeFileFlag::SmallFile){};
```
Comment (uncommon-nickname) [inline include/core/cfdp_core/pdu_directive.hpp:15]: Yeah, the same thing here - those two constructors are very weird

File: include/core/cfdp_core/pdu_directive.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Qlas) [reply include/core/cfdp_core/pdu_directive.hpp:94]: > isn't it plainly wrong?

Why?

[the remark this answers is not in the document]
Comment (Qlas) [reply include/core/cfdp_core/pdu_directive.hpp:94]: > Yeah, but this gives you this problem later...

At this point you should know if this is u32 or u64

[the remark this answers is not in the document]
Comment (Qlas) [reply include/core/cfdp_core/pdu_directive.hpp:94]: However, you're kind of right, it's a bit strange behavior

[the remark this answers is not in the document]
Comment (uncommon-nickname) [reply include/core/cfdp_core/pdu_directive.hpp:94]: > > isn't it plainly wrong?
> 
> Why?
Nvm, its fine

[the remark this answers is not in the document]
Comment (uncommon-nickname) [reply include/core/cfdp_core/pdu_directive.hpp:94]: > > Yeah, but this gives you this problem later...
> 
> At this point you should know if this is u32 or u64

Yeah, but you have to do an explicit cast every time you use the interface

File: src/cfdp_core/pdu_directive.cpp [hunk 6]
```diff
@@ -111,3 +113,119 @@ std::vector<uint8_t> cfdp::pdu::directive::AckPdu::encodeToBytes() const
...
+{
+    if (conditionCode == Condition::NoError)
+    {
+        throw exception::PduConstructionException(
+            "Fault location should be omitted with `No error` condition code");
+    }
+};
+
+cfdp::pdu::directive::EndOfFile::EndOfFile(std::span<uint8_t const> memory,
+                                           LargeFileFlag largeFileFlag, uint8_t lengthOfEntityID)
```
Comment (uncommon-nickname) [inline src/cfdp_core/pdu_directive.cpp:170]: how do we know in advance if incoming buffer contains a large or small file?

File: include/core/cfdp_core/pdu_directive.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (uncommon-nickname) [reply include/core/cfdp_core/pdu_directive.hpp:94]: Couldn't we template it somehow?

File: src/cfdp_core/pdu_directive.cpp (hunk 6 above)
Comment (Qlas) [reply src/cfdp_core/pdu_directive.cpp:170]: header has this values, we have to decode it first

File: src/cfdp_core/pdu_directive.cpp [hunk 7]
```diff
@@ -111,3 +113,119 @@ std::vector<uint8_t> cfdp::pdu::directive::AckPdu::encodeToBytes() const
...
+    const auto secondByte = memory[1];
+
+    conditionCode = Condition((secondByte & eof_condition_code_bitmask) >> 4);
+    checksum      = utils::bytesToInt<uint64_t>(memory, 2, sizeof(uint32_t));
+    fileSize      = utils::bytesToInt<uint64_t>(memory, 6, getSizeOfFileSize());
+
+    if (conditionCode != Condition::NoError)
+    {
+
+        if (memory[6 + getSizeOfFileSize()] != utils::toUnderlying(TLVType::EntityId))
```
Comment (uncommon-nickname) [inline src/cfdp_core/pdu_directive.cpp:193]: calculate this once

File: src/cfdp_core/pdu_directive.cpp [hunk 8]
```diff
@@ -111,3 +113,119 @@ std::vector<uint8_t> cfdp::pdu::directive::AckPdu::encodeToBytes() const
...
+    }
+};
+
+cfdp::pdu::directive::EndOfFile::EndOfFile(std::span<uint8_t const> memory,
+                                           LargeFileFlag largeFileFlag, uint8_t lengthOfEntityID)
+    : largeFileFlag(largeFileFlag), lengthOfEntityID(lengthOfEntityID)
+{
+    const auto memory_size = memory.size();
+
+    if (memory_size < const_small_file_pdu_size_bytes)
```
Comment (uncommon-nickname) [inline src/cfdp_core/pdu_directive.cpp:175]: what if its a large file? later you run direct memory access via `memory[6 + getSizeOfFileSize()]`. Wouldn't this crash if we got too small buffer?

File: src/cfdp_core/pdu_directive.cpp (hunk 6 above)
Comment (uncommon-nickname) [reply src/cfdp_core/pdu_directive.cpp:170]: Right

File: src/cfdp_core/pdu_directive.cpp (hunk 8 above)
Comment (Qlas) [reply src/cfdp_core/pdu_directive.cpp:175]: These calculations in all PDUs and header are pretty wrong. I will rethink this

Comment (uncommon-nickname) [reply src/cfdp_core/pdu_directive.cpp:175]: Why? Can you elaborate?

File: include/core/cfdp_core/pdu_directive.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Qlas) [reply include/core/cfdp_core/pdu_directive.hpp]: > Couldn't we template it somehow?

Added, but I don't like it, I will migrate it to uint64 + flag in constructor with condition in it.

File: src/cfdp_core/pdu_directive.cpp [hunk 9]
```diff
@@ -111,3 +113,119 @@ std::vector<uint8_t> cfdp::pdu::directive::AckPdu::encodeToBytes() const
 
     return encodedPdu;
 }
+
+cfdp::pdu::directive::EndOfFile::EndOfFile(Condition conditionCode, uint32_t checksum,
+                                           uint32_t fileSize)
+    : conditionCode(conditionCode), checksum(checksum), fileSize(fileSize),
+      largeFileFlag(header::LargeFileFlag::SmallFile)
+{
+    if (conditionCode != Condition::NoError)
```
[the remark this answers is not in the document]
Comment (Qlas) [reply src/cfdp_core/pdu_directive.cpp]: added isError method

File: include/core/cfdp_core/pdu_directive.hpp (hunk 5 above)
Comment (Qlas) [reply include/core/cfdp_core/pdu_directive.hpp:15]: SAA

File: src/cfdp_core/pdu_directive.cpp (hunk 8 above)
Comment (Qlas) [reply src/cfdp_core/pdu_directive.cpp]: I checked they again and they look good.
I had to add additional size validation after we know condition

File: src/cfdp_core/pdu_directive.cpp [hunk 10]
```diff
@@ -111,3 +113,107 @@ std::vector<uint8_t> cfdp::pdu::directive::AckPdu::encodeToBytes() const
...
+
+    if (not isError())
+    {
+        return encodedPdu;
+    }
+
+    encodedPdu.push_back(utils::toUnderlying(TLVType::EntityId));
+
+    auto faultEntityIDBytes = utils::intToBytes(faultEntityID, lengthOfEntityID);
+    utils::concatenateVectorsInplace(faultEntityIDBytes, encodedPdu);
```
Comment (Qlas) [inline src/cfdp_core/pdu_directive.cpp:213]: TLV format is wrong
image

Comment (Qlas) [reply src/cfdp_core/pdu_directive.cpp:213]: fixed

File: include/core/cfdp_core/pdu_directive.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Qlas) [reply include/core/cfdp_core/pdu_directive.hpp]: migrated to uin64 with flag

File: tests/cfdp_core/pdu_directive.test.cpp (hunk 3 above)
Comment (Qlas) [reply tests/cfdp_core/pdu_directive.test.cpp]: done

File: include/core/cfdp_core/pdu_directive.hpp (hunk 5 above)
Comment (Qlas) [reply include/core/cfdp_core/pdu_directive.hpp]: agree

**Review by uncommon-nickname — APPROVED** [review]
LGTM

**Pull request merged**